dagstore_control.go

```diff
@@ -14,6 +14,8 @@ const (
 	OpShardAcquire
 	OpShardFail
 	OpShardRelease
+
+	opFlush // only for tests
```

Comment:
This pattern is racy and leaky. And it's not really a flush because it's not technically _flushing_ anything. It's really inserting a "bookmark/breakpoint" in the event queue to get notified when it's hit. The problem is that this solution assuming that all events of interest have been executed _before_ the mark is inserted. That's why it's racy (they may not have), and leaky (as a developer, you need to know enough of the internals to know that everything interesting has executed before you insert the mark).

What you're trying to solve here in a low-key fashion is the lack of observability, which is useful both for production usage and testing.

Let's instead build a "notifier" mechanism, which you subscribe to from tests, but that can also be used by users to watch shard state transitions.

Reply:
Makes a lot of sense.